src/utils/file.cpp

```diff
@@ -233,27 +276,139 @@ namespace file_util {
     return ret;
   }
 
+  /*
+   * Return the path without the last component (see dirname(3))
+   */
+  string dirname(const string& path) {
+    if (path.empty()) {
+      return ".";
+    }
+
+    auto start = path.size() - 1;
+    if (path[start] == '/' && start > 0) {
+      --start;
+    }
+
+    auto pos = path.rfind('/', start);
+    if (pos == string::npos) {
+      return ".";
+    }
+
+    if (pos == 0) { // path == "/" or path == "/directory"
+      return "/";
+    }
+
+    return path.substr(0, pos);
+  }
+
   /**
    * Path expansion
    */
-  const string expand(const string& path) {
-    string ret;
-    vector<string> p_exploded = string_util::split(path, '/');
-    for (auto& section : p_exploded) {
-      switch(section[0]) {
-        case '$':
-          section = env_util::get(section.substr(1));
-          break;
-        case '~':
-          section = env_util::get("HOME");
-          break;
+  string expand(const string& path) {
+    if (path.empty()) {
+      return "/";
+    }
+
+    auto ret = path;
+    if (ret[0] == '~') {
+      file_util::expand_home_dir(ret);
+    }
+
+    vector<string> p_exploded = string_util::split(ret, '/');
+    for (auto it = p_exploded.begin(); it != p_exploded.end(); ) {
+      if (it->empty() || *it == ".") {
+        it = p_exploded.erase(it);
+      } else if (*it == "..") {
+        if (it != p_exploded.begin()) {
+          --it;
+          it = p_exploded.erase(it);
+        }
+
+        it = p_exploded.erase(it);
+      } else {
+        auto& path = *it;
```

Comment:
This is confusing because it has the same name as the parameter

Reply:
Good catch, changed it to "component"